Ai sdk 5 migration

timtmok opened the pull request:
Address #10818 

This used Vercel's MCP server to assist in the migration to v5.

* Removed the workaround for the temperature
* Updated the field names that changed
* Fixed the responses to match the v5 format

### Release Notes

#### New Features

- Update `ai-sdk` to v5

#### Bug Fixes

- N/A


### QA Notes
I did some light testing with tool calling with various providers. The chat response handling is where things changed the most.

timtmok committed:
Update packages to AI SDK 5

timtmok committed:
Fix up tool calls

sharon-wang reviewed (changes requested):
Looking good so far with Amazon Bedrock, OpenAI and OpenRouter via Custom Provider 👍 getPlot tool is working with Amazon Bedrock, so I wonder if our plot image handling is compatible with Anthropic models but not the OpenAI models?

Providers that don't support the responses endpoint aren't working though -- I think we'll need some way to override the chat endpoint for those providers to `/chat/completions`

timtmok committed:
Fix Snowflake provider

timtmok commented on extensions/positron-assistant/src/openai-fetch-utils.ts:
I kept this in for Snowflake compatibility. I was getting errors without this handling.

wch commented on extensions/positron-assistant/src/utils.ts:
```diff
@@ -212,46 +232,64 @@ export function markBedrockCacheBreakpoint(message: ai.CoreMessage): ai.CoreMess
 function convertToolResultToAiMessageExperimentalContent(
```
This function name (and others with `ExperimentalContent`) is a bit of a misnomer, now that the output message no longer has an `experimental_content` field.

wch commented on extensions/positron-assistant/src/utils.ts:
```diff
@@ -260,22 +298,22 @@ function convertToolResultToAiMessageExperimentalContent(
+ * AI SDK 5 supports images in tool results via the 'content' output type with 'media' parts.
  */
-function getPlotToolResultToAiMessage(part: vscode.LanguageModelToolResultPart2): ai.CoreUserMessage {
+function getPlotToolResultToAiMessage(part: vscode.LanguageModelToolResultPart2): ai.ToolModelMessage {
```
Just wondering: is this function still needed? I think that `convertToolResultToAiMessageExperimentalContent` should be able to handle tool results with images.

wch commented on extensions/positron-assistant/src/utils.ts:
This might be outside the scope of this PR, but does this only mark cache breakpoints with Bedrock? What about Anthropic (not via Bedrock)?

wch commented on extensions/positron-assistant/src/utils.ts:
```diff
@@ -89,17 +92,34 @@ export function toAIMessage(
+								.join('');
+
+							// Try to parse as JSON if it looks like JSON, otherwise use as text
+							let output: { type: 'json'; value: unknown } | { type: 'text'; value: string };
```
There is a cast down below which can be removed if this type is changed:

```suggestion
							let output: { type: 'json'; value: ai.JSONValue } | { type: 'text'; value: string };
```

wch commented on extensions/positron-assistant/src/utils.ts:
```diff
@@ -212,46 +232,64 @@ export function markBedrockCacheBreakpoint(message: ai.CoreMessage): ai.CoreMess
+		// Convert text parts to string for output
+		// AI SDK 5 expects { type: 'text', value: string }
+		const textContent = part.content
+			.filter((c): c is vscode.LanguageModelTextPart => c instanceof vscode.LanguageModelTextPart)
```
```suggestion
```

wch commented on extensions/positron-assistant/src/openai-fetch-utils.ts:
```diff
@@ -124,8 +127,42 @@ function transformServerSentEvents(text: string, providerName: string): string {
 				// Fix empty role fields in delta objects within choices array
 				if (data.choices && Array.isArray(data.choices)) {
```
Using the type guard function from above:

```suggestion
				if (isChatCompletionChunk(data)) {
```

(Although, do Snowflake and others provide chunks that don't quite fit this format?)

wch commented on extensions/positron-assistant/src/openai-fetch-utils.ts:
```diff
@@ -124,8 +127,42 @@ function transformServerSentEvents(text: string, providerName: string): string {
 					for (const choice of data.choices) {
-						if (choice.delta && typeof choice.delta === 'object' && choice.delta.role === '') {
-							choice.delta.role = 'assistant';
+						if (choice.delta && typeof choice.delta === 'object') {
```
With the type check, this `if` statement is no longer necessary.

```suggestion
```

wch commented on extensions/positron-assistant/src/openai-fetch-utils.ts:
Using the types from `OpenAI`, I suggest actually extracting a lot of the code above into a function which takes as input something that's close to a `ChatCompletionChunk`, but where the type takes into account the possibly-missing fields. The function would return something that actually is a `ChatCompletionChunk`.

Then you can really lean into type checking to make sure all the cases are taken care of. 


For example, if you define this type, then the `role` could be `""`:

```ts
type PossiblyBrokenChatCompletionChunk = Omit<OpenAI.ChatCompletionChunk, 'choices'> & {
	choices: Array<Omit<OpenAI.ChatCompletionChunk['choices'][0], 'delta'> & {
		delta: Omit<OpenAI.ChatCompletionChunk['choices'][0]['delta'], 'role'> & {
			role?: OpenAI.ChatCompletionChunk['choices'][0]['delta']['role'] | ''
		}
	}>
};
```

I know that's kind of awkward but it does work. It could probably be made a bit simpler and easier to read by referencing the types by name instead of by path. (I would just ask an LLM to add in the other possibly-empty fields as well.)

Next, you'd modify the type guard function above to have the signature:

```ts
export function isPossiblyBrokenChatCompletionChunk(obj: unknown): obj is PossiblyBrokenChatCompletionChunk {
    // body of this function can remain the same
}
```

Then you could add a function with this signature:

```ts
function fixPossiblyBrokenChatCompletionChunk(chunk: PossiblyBrokenChatCompletionChunk): OpenAI.ChatCompletionChunk {
	// Implementation goes here
}
```


Then up above you'd use:

```ts
if (isPossiblyBrokenChatCompletionChunk(data)) {
  const fixedChunk = fixPossiblyBrokenChatCompletionChunk(data)
  transformedLines.push(`data: ${JSON.stringify(fixedChunk)}`);
```

sharon-wang reviewed (commented):
I can log in with Snowflake now ❄️ ✅ 

I'm not sure if this is my local state, but I lost the Snowflake icon in the models popup:

<img width="758" height="454" alt="Image" src="https://github.com/user-attachments/assets/5c9cfcb2-b637-41bd-98cd-21a6ac7ed991" />

The icon is there in a release build of 2026.01.0+24, so not sure if it's my dev build state or possibly something has changed?

timtmok committed:
Clean up types and experimental content

timtmok committed:
Revert test code

timtmok commented:
I tried these latest changes with different models and providers and things feel good. The bonus is Snowflake works with GPT models calling the GetPlot tool. Its response is brief but it is able to understand the image.

<img width="296" height="301" alt="image" src="https://github.com/user-attachments/assets/7d8f0b92-7090-46da-838d-aba02f06f1cb" />

The response is from GPT-5 but was also similar with GPT-4.1.

<img width="272" height="117" alt="image" src="https://github.com/user-attachments/assets/4184d28b-b459-4a23-b7f9-6ec0c144e622" />

The Snowflake icon seems okay for me as well but these changes shouldn't have changed that.

wch commented on extensions/positron-assistant/src/openai-fetch-utils.ts:
```diff
@@ -120,17 +266,15 @@ function transformServerSentEvents(text: string, providerName: string): string {
-					}
+				// Check if it's a possibly broken chunk and fix it
+				// Otherwise, keep the original line
+				if (isPossiblyBrokenChatCompletionChunk(data)) {
```
Just a follow on to the note above about the type and type guard mismatch: if `isPossiblyBrokenChatCompletionChunk()` actually matched the `PossiblyBrokenChatCompletionChunk` type, then it would always enter this `if` block and never go into the `else`. (Assuming that `data` is a JS object, as opposed to something like a string or null.)

sharon-wang reviewed (commented):
General functionality looking good! I'm seeing some issues with tool calling still:

- we can address the getPlot issues separately since we've got other issues tracking that
- the project tree tool and text search tool (and maybe any tools with required input params?) are showing some odd behaviour for Snowflake though. I wonder if it has to do with removing "strict"?

timtmok committed:
Fix Snowflake tool call error

timtmok committed:
Fix OpenAI compatible providers tools

Tools with no required args need a fix so that the tool will be called properly

timtmok commented:
I ended up going with a simpler check to see if the data is a chat completion chunk. It was simpler to see if it says it's a chunk then go into all the fields to see if they need any fixing. It got to complicated checking if any fields were broken and that code became just as long as fixing it.

wch committed:
Simplify code to fix PossiblyBrokenChatCompletionChunk

timtmok committed:
Export type for test

timtmok committed:
Fix other errors with openai compatible providers

sharon-wang reviewed (commented):
Looking good and working well for me!

get plot related issues (for Snowflake and OpenRouter + Custom Provider) will be addressed separately in https://github.com/posit-dev/positron/issues/10759

Looks like the Snowflake test might need some updates:

```
  1) Snowflake Provider
       OpenAI Compatible Fetch - Snowflake Streaming Fix
         transforms empty role fields in streaming response:

      AssertionError [ERR_ASSERTION]: Empty role should be transformed to "assistant"
      + expected - actual

      -false
      +true
```

timtmok committed:
Merge branch 'main' into ai-sdk-5-migration

timtmok committed:
Fix test and move openai dependency

timtmok committed:
Merge branch 'main' into ai-sdk-5-migration

sharon-wang reviewed (commented):
Looking good! One main thing to fix up for snowflake, and then I think let's get this in :shipit: 

### Snowflake Cortex

- I think we may have lost the handling to use /chat/completions instead of /responses, since I'm seeing an error in chat for any message, and seeing the following in the logs

#### error in chat

```js
[Claude Sonnet 4.5] Error in chat response: { "name": "AI_APICallError", "url": "https://<ACCOUNT_ID>.snowflakecomputing.com/api/v2/cortex/v1/responses"
```

#### error in output log

```js
2026-01-19 11:07:48.343 [debug] [Snowflake Cortex] [DEBUG] Response status: 404 Not Found
```

---

### getPlot tool

I think we're at a reasonable spot with Custom Provider and can fix up the plot tool issue in https://github.com/posit-dev/positron/issues/10759. Documenting my test steps for reference:

#### Steps

1. Run [plot-hallucination.r](https://github.com/posit-dev/qa-example-content/blob/main/workspaces/assistant/plot-hallucination.r) in the Console
2. Do not attach the script as context
3. In Agent mode, ask "use the getplot tool, then tell me what is the number in the middle of the plot? what colour is it?"
 
#### Results

| Provider | Model | Result | 📸 |
|--------|--------|--------|--------|
| Custom Provider (OpenRouter) | anthropic/claude-sonnet-4 | ❌ hallucinated number and colour | <img width="958" height="773" alt="Image" src="https://github.com/user-attachments/assets/f3d1b2ee-8a48-4629-a57a-77d83dd9ecca"/> |
| Custom Provider (OpenRouter) | openai/gpt-4.1 | ❌ hallucinated number and colour | <img width="956" height="770" alt="Image" src="https://github.com/user-attachments/assets/057ce472-dbcb-4e01-8ae9-03c31ceb9d34" /> |
| OpenAI | gpt-4.1 | ✅ correct number and colour | <img width="957" height="775" alt="Image" src="https://github.com/user-attachments/assets/c51566c4-7cf5-4f6a-980d-34e4b115aa7c" /> |
| OpenAI | gpt-5.2 | ✅ correct number and colour | <img width="956" height="771" alt="Image" src="https://github.com/user-attachments/assets/fdc95469-55e2-4bec-ab8c-50e470e229cc" /> |

timtmok committed:
Merge branch 'main' into ai-sdk-5-migration

timtmok committed:
Fix Snowflake provider

timtmok committed:
Fix connections.toml usage

sharon-wang reviewed (approved):
LGTM!
